overlord/devicestate/devicestate.go

```diff
@@ -1411,35 +1442,289 @@ func createRecoverySystemTasks(st *state.State, label string, snapSetupTasks []s
+		// revisions.
+		ts, info, err := snapstateDownload(context.TODO(), st, sn.Name, dirs.SnapBlobDir, &snapstate.RevisionOptions{
+			Channel:        sn.DefaultChannel,
+			Revision:       rev,
+			ValidationSets: valsets.Keys(),
+		}, userID, snapstate.Flags{}, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		tracker.Add(info)
+		downloadTSS = append(downloadTSS, ts)
+	}
+
+	warnings, errs := tracker.Check()
+	for _, w := range warnings {
+		logger.Noticef("create recovery system prerequisites warning: %v", w)
+	}
+
+	// TODO: use function from other branch
+	if len(errs) > 0 {
+		var builder strings.Builder
+		builder.WriteString("cannot create recovery system from model that is not self-contained:")
+
+		for _, err := range errs {
+			builder.WriteString("\n  - ")
+			builder.WriteString(err.Error())
+		}
+
+		return nil, errors.New(builder.String())
+	}
+
+	var snapsupTaskIDs []string
+	if len(downloadTSS) > 0 {
+		snapsupTaskIDs, err = extractSnapSetupTaskIDs(downloadTSS)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	chg = st.NewChange("create-recovery-system", fmt.Sprintf("Create new recovery system with label %q", label))
+	createTS, err := createRecoverySystemTasks(st, label, snapsupTaskIDs, opts)
 	if err != nil {
 		return nil, err
 	}
-	chg.AddAll(ts)
+
+	chg.AddAll(createTS)
+
+	for _, ts := range downloadTSS {
+		createTS.WaitAll(ts)
+		chg.AddAll(ts)
+	}
+
 	return chg, nil
 }
 
+func checkForSnapIDs(model *asserts.Model, localSnaps []LocalSnap) error {
+	for _, sn := range model.AllSnaps() {
+		if sn.ID() == "" {
+			return fmt.Errorf("cannot create recovery system from model with snap that has no id: %q", sn.Name)
+		}
+	}
+
+	for _, sn := range localSnaps {
+		if sn.SideInfo.SnapID == "" {
+			return fmt.Errorf("cannot create recovery system from provided snap that has no id: %q", sn.SideInfo.RealName)
+		}
+	}
+
+	return nil
+}
+
+func offlineSnapInfo(sn *asserts.ModelSnap, rev snap.Revision, opts CreateRecoverySystemOptions) (*snap.Info, error) {
+	index := -1
+	for i, si := range opts.LocalSnaps {
+		if sn.ID() == si.SideInfo.SnapID {
+			index = i
+			break
+		}
+	}
+	if index == -1 {
+		return nil, fmt.Errorf(
+			"missing snap from local snaps provided for offline creation of recovery system: %q, rev %v", sn.Name, rev,
+		)
+	}
+
+	localSnap := opts.LocalSnaps[index]
+
+	if !rev.Unset() && rev != localSnap.SideInfo.Revision {
+		return nil, fmt.Errorf(
+			"snap %q does not match revision required by validation sets: %v != %v", localSnap.SideInfo.RealName, localSnap.SideInfo.Revision, rev,
+		)
+	}
+
+	s, err := snapfile.Open(localSnap.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	return snap.ReadInfoFromSnapFile(s, localSnap.SideInfo)
+}
+
+func snapNeedsInstall(st *state.State, name string, rev snap.Revision) (bool, error) {
+	info, err := snapstate.CurrentInfo(st, name)
+	if err != nil {
+		if isNotInstalled(err) {
+			return true, nil
+		}
+		return false, err
```

Comment:
we don't seem to test this case

Reply:
I'm not actually sure how to test that case in a meaningful way. What would cause `snapstate.CurrentInfo` to return something that isn't a `snap.NotInstalledError`?

Reply:
read issues reading from snap, anyway its probably fine without test, I think I somehow read it as `false, nil` instead of `false, err`